packages/blocks/plus_minus/src/index.js

```diff
@@ -0,0 +1,1048 @@
+      this.removePart_();
+    }
+    this.updateMinus_();
+  },
+
+  /**
+   * Callback for the plus image. Adds an input to the block and updates the
+   * state of the minus.
+   * @this Blockly.Block
+   */
+  plus: function() {
+    this.addPart_();
+    this.updateMinus_();
+  },
+
+  /**
+   * Callback for the minus image. Removes the input at the end of the block and
+   * updates the state of the minus.
+   * @this Blockly.Block
+   */
+  minus: function() {
+    if (this.itemCount_ == 0) {
+      return;
+    }
+    this.removePart_();
+    this.updateMinus_();
+  },
+
+  // To properly keep track of indices we have to increment before/after adding
+  // the inputs, and decrement the opposite.
+  // Because we want our first item to be ADD0 (not ADD1) we increment after.
+
+  /**
+   * Adds an input to the end of the block. If the block currently has no
+   * inputs it updates the top 'EMPTY' input to receive a block.
+   * @this Blockly.Block
+   * @private
+   */
+  addPart_: function() {
+    if (this.itemCount_ == 0) {
+      if (this.getInput('EMPTY')) {
+        // TODO: I don't think this should throw errors. It would be nice if
+        //   It returned a boolean instead.
+        this.removeInput('EMPTY');
+      }
+      this.topInput_ = this.appendValueInput('ADD' + this.itemCount_)
+          .appendField(new FieldPlus(), 'PLUS')
+          .appendField(Blockly.Msg['TEXT_JOIN_TITLE_CREATEWITH']);
+    } else {
+      this.appendValueInput('ADD' + this.itemCount_);
+    }
+    this.itemCount_++;
+  },
+
+  /**
+   * Removes an input from the end of the block. If we are removing the last
+   * input this updates the block to have an 'EMPTY' top input.
+   * @this Blockly.Block
+   * @private
+   */
+  removePart_: function() {
+    this.itemCount_--;
+    this.removeInput('ADD' + this.itemCount_);
+    if (this.itemCount_ == 0) {
+      this.topInput_ = this.appendDummyInput('EMPTY')
+          .appendField(new FieldPlus(), 'PLUS')
+          .appendField(this.newQuote_(true))
+          .appendField(this.newQuote_(false));
+    }
+  },
+
+  /**
+   * Makes it so the minus is visible iff there is an input available to remove.
+   * @private
+   */
+  updateMinus_: function() {
+    const minusField = this.getField('MINUS');
+    if (!minusField && this.itemCount_ > 0) {
+      this.topInput_.insertFieldAt(1, new FieldMinus(), 'MINUS');
+    } else if (minusField && this.itemCount_ < 1) {
+      this.topInput_.removeField('MINUS');
+    }
+  },
+};
+
+/**
+ * Adds the quotes mixin to the block. Also updates the shape so that if no
+ * mutator is provided the block has two inputs.
+ * @this Blockly.Block
+ * @constructor
+ */
+const textJoinHelper = function() {
+  this.mixin(Blockly.Constants.Text.QUOTE_IMAGE_MIXIN);
+  this.updateShape_(2);
+};
+
+Blockly.Extensions.unregister('text_join_mutator');
+Blockly.Extensions.registerMutator('text_join_mutator',
+    textJoinMutator, textJoinHelper);
+
+const listCreateMutator = {
```

Comment:
Note: this mutator is very similar to the previous one, differing mostly in the `addPart_` and `removePart_` functions.

Reply:
I'm thinking of ways to combine them but they all add another layer of abstraction and I'm not convinced that it's work adding that to avoid the code duplication.

@samelhusseini thoughts?  Look at `mutationToDom`, `domToMutation`, `updateShape_`, `plus`, `minus`, and `updateMinus_` on this and on the previous chunk.

Reply:
[This](http://bekawestberg.me/blog/mutator-api) might be of interest. I'm not super attached to the idea, but I think it would be one way to go if you're interested in API/reusability stuffs.

Basic idea is to abstract mutators into objects (think strategy pattern style interchangability). Users would have to pass in a top input, extra input, and empty input to initialize the mutator.